File: 02RubyChatServer/chatserver.rb
require "socket"
require "thread"

class ChatServer
    def initialize(port)
        @clients = Array::new
        @clientSockets = Array::new
        @serverSocket = TCPServer.new("", port)
        @serverSocket.setsockopt(Socket::SOL_SOCKET, Socket::SO_REUSEADDR, 1)
        printf("ChatServer started on port %d\n", port)
    end
    def run
        while 1
            Thread.start(@serverSocket.accept) do |clientSocket|
                newClient = ChatClient.new(clientSocket, self)
                printf("New Client %s:%s\n",newClient.clientSocket.peeraddr[2], newClient.clientSocket.peeraddr[1])
                @clients.push(newClient)
                begin
                    newClient.run
                rescue Errno::ECONNRESET => e
                    printf("Client Left %s:%s\n",newClient.clientSocket.peeraddr[2], newClient.clientSocket.peeraddr[1])
                    newClient.clientSocket.close
                    @clients.delete(newClient)
                    Thread.current.kill
                end
            end
        end
    end

    def messageController(message)
        messageToEverybody(message)
    end

    def messageToEverybody(message)
        for client in @clients do
            begin
                client.messageToClient(message) 
            rescue => exception
                puts exception
            end
        end        
    end
end

class ChatClient
    def initialize(clientSocket, server)
        @clientSocket = clientSocket
        @server = server
    end

    def messageToClient(message)
        @clientSocket.puts(message)
    end

    def run
        while 1
            message = @clientSocket.gets()
            @server.messageController(message)
        end
    end
    def clientSocket
        @clientSocket
    end
end

myChatServer = ChatServer.new(5000)
myChatServer.run

File: Flask/02PythonChatServer/chatserver.py
#!/usr/bin/env python3
from socket import AF_INET, socket, SOCK_STREAM
from threading import Thread
BUFSIZ = 1024

class ChatServer:
  def __init__(self, host, port):
    self.__host = host
    self.__port = port
    self.__clients = []
    self.__serverSocket = socket(AF_INET, SOCK_STREAM)
    self.__serverSocket.bind((host, port))

  def run(self):
    self.__serverSocket.listen()
    print(f"ChatServer started on {self.__host}:{self.__port}\n")
    while True:
      clientSocket, clientAddress = self.__serverSocket.accept()
      print("%s:%s has connected." % clientAddress)
      newClient = ChatClient(clientSocket, self)
      self.__clients.append(newClient)
      Thread(target=newClient.run).start()

  def messageController(self, message):
    self.messageToEverybody(message)
  
  def messageToEverybody(self, message):
    for client in self.__clients:
      client.messageToClient(message)

  def removeClient(self, client):
    client.getClientSocket().close()
    clients.remove(client)

  def close(self):
    self.__serverSocket.close()

class ChatClient:
  def __init__(self, clientSocket, chatServer):
    self.__clientSockets = clientSocket
    self.__chatServer = chatServer

  def messageToClient(self, message):
    self.__clientSockets.send(message)
  
  def run(self):
    while True:
      try:
        message = self.__clientSockets.recv(BUFSIZ)
        self.__chatServer.messageController(message)
      except ConnectionResetError:
        self.__chatServer.removeClient(client)
  
  def getClientSocket(self):
    return self.__clientSockets()

if __name__ == "__main__":
  chatServer = ChatServer('', 5000)
  ACCEPT_THREAD = Thread(target=chatServer.run)
  ACCEPT_THREAD.start()
  ACCEPT_THREAD.join()
  chatServer.close()
